### ISSUE TITLE
[RFC] Proposal for Reference Value Provider Service

### ISSUE THREAD

**Issue by Xynnn007** [confidential-containers/confidential-containers#122]
# Summary
This proposal describes Reference Value Provider Service, a key component to generating reference value, which handles different provenances, extracts reference value, and sends it to Attestation Service.

# Motivation
![image](https://user-images.githubusercontent.com/22250460/173333816-de0c0d39-72b4-4e7c-8c80-661345b6cc2a.png)
In confidential container scenario, the containers are running in TEE. However, other privileged software components also running in the same TEE, like guest OS, kernel, firmware, kata-agent, etc., can eavesdrop and tamper with information of containers. Thus, the `verifier`(e.g., tenant) must verify the identities of the components running in the TEE, ensuring that TCB status is as expected.

The way of the verification is to match evidence from attester and reference value from Reference Value Provider. The reference value may including two parts. One is hardware-specific part, used to check whether the hardware is working as expected. The other is some claims, the claims may including the digest of the trusted components (We say `trusted digests`). By comparation between gathered unverified digests and trusted digests, `verifier` can know whether he/she can trust the the privileged components. 

This proposal only aims for trusted digests. How other parts comes, like the hardware-specific part, is out of scope in this proposal. 

A trust digest may origin from different sources:

|Trust Model| Strategy| Examples|
|:-:| :-:| :-:| 
|The `verifier` trusts himself| Directly gets digest from the `verifier`.| Manually input |
|The `verifier`  trusts the signer of `provenance` (3rd party)| Gets, verifies，and extracts trusted digest from the `provenance` | *VSA* in SLSA、*in-toto* in CNCF、*CoRIM* and *CoSWID* in IETF RATS workgroup |
|`verifier` trusts the source code of the artifact (code transparency)| Gets `provenance`，build the binary(artifact) according to scripts or commands proveded in the `provenance`, and then get the digest. | *Reproducible-build* |

The term `provenance` refers to signed metadata which records the build pipeline or supply chain steps of the artifact, usually containing the artifact's id, digest, owner, etc., always signed and issued by a third entity that is usually the vendor of the artifact. It implies that the `verifier` trusts the third party. Also, the term `provenance` can refer to some guide (e.g., build scripts and source code repo) for reproducible build of the artifact. All the "provenance information" of this artifact can be treated as `provenance`.

![image](https://user-images.githubusercontent.com/22250460/173334685-dfe7a861-ddaf-4dea-ba38-3f03552cf74e.png)

To make it easier for Attestation Service to receive trusted digests in a uniform layout, we need a component to verify different `provenance`s and extract trusted digests. This component is Reference Value Provider Service (RVPS).

![image](https://user-images.githubusercontent.com/22250460/173334734-782f8a54-ba64-4c24-8512-73ae101a8731.png)

Generally, a `verifier` can implement a completely customized RVPS according to their specific requirements. But the CC community still lacks a reference design for RVPS. The lack of reference design causes confusion and inconvenience for potential CC users to use this confidential container solution.

# Background
Here is the proposed architecture Verified Build Distribution Architecture (VBDA). VBDA can show the position of RVPS in CC environment.
![image](https://user-images.githubusercontent.com/22250460/173334824-678f32cf-dcbe-437f-b44d-bbe439b79a29.png)
VBDA gives a sketch of a lifetime of an artifact, from coding to deployment in TEE. Binary Provenance Distribution Service stores artifacts and their `provenance`. When preparing to deploy artifacts (i.e., system software), CSP can directly get the artifacts from BPDS.

# Short Term Goals
* Support different formats of provenances by modulization.
* Distribute the RVPS as a library or single service.
* Implement in-toto v0.9 as a kind of provenance.

# Long Term Goals
* Define the protocol between BPDS and RVPS
* Support more provenance types (CoRIM, SLSA VSA, SLSA Provenance, etc.)

# Architecture
## As a single executable

![image](https://user-images.githubusercontent.com/22250460/178392607-c6611e2b-5490-4176-bdde-f8975ef0244e.png)

Here is the whole architecture of RVPS. There are four parts:
* Provenance API: Interface that receives `provenance`s from BPDS or tenant.
* Pre-Processor: Pre-process the `provenance` received by Provenance API. The RVPS's owner can add customized wares to Pre-Processor to process the `provenance`s.
* Extractors: Core of RVPS, including provenance-specific sub-modules, which verify `provenance` and extract digests from `provenance`.
* Cache: Storage for trusted digests.
* AS-API: Module responsible for communication with Attestation Service.

## As a submodule in AS

![image](https://user-images.githubusercontent.com/22250460/178392741-f193c617-be46-4967-b404-1ac991936e09.png)

If RVPS a submodule is AS, provenance messages need to be received by Attestation Service rather than RVPS itself.
 
### Data Flow as a submodule
![image](https://user-images.githubusercontent.com/22250460/178393346-f8175062-702e-4436-a3af-6865523f35da.png)

### Provenance API
Responsible for receiving provenance.

### Pre-Processor
A set of `Ware`s, which processes message before the message is delivered to Extractors.
The top level interface of PreProcessor:

```rust
pub trait PreProcessorAPI {
    fn process(&self, message: &mut Message) -> Result<()>;
    fn add_ware(&mut self, ware: Box<dyn Ware>) -> &Self;
}
```

Here, `Ware` is a trait for all `Ware`s
```rust
/// A Ware loaded in Pre-Processor will process all the messages passing
/// through the Pre-Processor. A series of Wares organized in order can
/// process all the messages in need before they are consumed by the
/// Extractors.
pub trait Ware {
    fn handle(
        &self,
        message: &mut Message,
        context: &mut HashMap<String, String>,
        next: Next<'_>,
    ) -> Result<()>;
}

/// Next encapsulates the remaining ware chain to run in [`Ware::handle`].
```

Here
* `context` conveys context variables between `Ware`s
* `message` is the message being processed by the RVPS. This message is a wrapper for provenances. Refer to [message layout](https://github.com/Xynnn007/reference-value-provider-service/blob/main/docs/IMPLEMENTATION.md#message)

### Extractors
Extractors contain different `Extractor`, which verifies and extracts trusted digests from different provenances.
Each `Extractor` must implement the following trait
```rust
pub trait Extractor {
    fn process(&mut self, message: Message) -> Result<ReferenceValue>;
}
```
* `Message` is a defined universal format for provenance. Like `GetResourceRequest` Message in `Attestation Agent`'s app implementation.
In POC, an `Extractor` supporting *in-toto* will be implemented.

### Cache
A `Cache` is a storage module to store trusted digests, which is helpful to restore data after RVPS restarts or crashes. A `Cache` must implement the following trait
```rust
trait Cache {
    fn Put(&mut self, artifact_name: String, reference_value: ReferenceValue) -> Result<()>;
    fn Get(artifact_name: String, &self) -> Result<ReferenceValue>;
    // fn Revoke(&mut self, artifact_name: String) -> Result<()>;
}
```

# Reference
## in-toto
*in-toto* is a framework to secure the integrity of software's supply chain, which is in CNCF now. It can produce a provenance of an artifact. Its trust model is as the figure
![image](https://user-images.githubusercontent.com/22250460/173336117-5d546439-c452-404f-bb2e-8393c56d8eed.png)

When trusting the Project Owner (who owns the supply chain of the artifact), RVPS can verify and extract trusted digest from the provenance of the artifact.

## SLSA Provenance & VSA
SLSA is a security framework and a checklist of standards for software artifacts. It can produce provenances named *Provenance* and *VSA*. The new version of *in-toto* in development will support SLSA. The basic trust model is similar to *in-toto*, requiring the `verifier` to trust the signers of *Provenance* and *VSA* to verify the validity and extract trusted digests.

## CoRIM 
*CoRIM* is a reference value format for describing firmware defined by the IETF RATS working group. The basic trust model is that the verifier must trust the signer of *CoRIM* so that they can verify the validity of *CoRIM* and extract the relevant trusted digest.

# Links
https://www.ietf.org/archive/id/draft-ietf-rats-architecture-15.txt
https://www.ietf.org/archive/id/draft-birkholz-rats-corim-02.html
https://www.ietf.org/archive/id/draft-ietf-sacm-coswid-20.txt
https://github.com/in-toto/docs/blob/v0.9/in-toto-spec.md
https://in-toto.io/
https://slsa.dev/verification_summary/v0.1
https://slsa.dev/provenance/v0.2

**Comment by Xynnn007** [comment]
> I was looking to see if Trustee had some work done on SLSA provenance verification, would this be it?

Yes, partially.

The core goal of RVPS is to "obtain reference values ​​from a trusted supply chain." And the SLSA verification you mentioned is precisely that "trusted" aspect.

**Comment by nmwael** [comment]
Okay, it should be somewhat "simple" to add an rvps that could support json based provenance, im not sure though how much it would take on the AA side though.. Potentially it could be detached so the provenance were verified at the rootfs and then only the provenance json had to be measured across to keep the overhead down.

**Comment by nmwael** [comment]
Yes, this was what I was thinking about although specifically for the software parts..
